lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -3670,40 +3670,114 @@ void TypeChecker::checkParameterList(ParameterList *params,
   }
 }
 
-ArrayRef<Decl *>
+static bool addSpecializationConstraint(
+    constraints::ConstraintLocator *locator, Type boundType,
+    ArrayRef<TypeRepr *> specializationArgs,
+    DeclContext *dc, constraints::ConstraintSystem &cs) {
+  using namespace constraints;
+  // Resolve each type.
+  SmallVector<Type, 2> specializationArgTypes;
+  auto options =
+      TypeResolutionOptions(TypeResolverContext::InExpression);
+  for (auto specializationArg : specializationArgs) {
+    const auto result = TypeResolution::resolveContextualType(
+        specializationArg, dc, options,
+        // Introduce type variables for unbound generics.
+        OpenUnboundGenericType(cs, locator),
+        HandlePlaceholderType(cs, locator),
+        OpenPackElementType(cs, locator, nullptr));
+    if (result->hasError())
+      return true;
+
+    specializationArgTypes.push_back(result);
+  }
+  cs.addConstraint(
+      ConstraintKind::ExplicitGenericArguments, boundType,
+      PackType::get(cs.getASTContext(), specializationArgTypes),
+      locator);
+  return false;
+}
+
+BraceStmt *
 ExpandMacroExpansionDeclRequest::evaluate(Evaluator &evaluator,
                                           MacroExpansionDecl *MED) const {
   auto &ctx = MED->getASTContext();
   auto *dc = MED->getDeclContext();
-  auto foundMacros = TypeChecker::lookupMacros(
-      MED->getDeclContext(), MED->getMacro(),
-      MED->getLoc(), MacroRole::Declaration);
+  // Non-call declaration macros cannot be overloaded.
+  MacroRoles viableRoles(MacroRole::Declaration);
+  // In a code block context, a macro expansion decl can also expand to an
+  // arbitrary code item.
+  if (dc->isLocalContext() ||
+      dc->getContextKind() == DeclContextKind::FileUnit) {
+    viableRoles |= MacroRole::Expression;
+    viableRoles |= MacroRole::CodeItem;
+  }
```

Comment:
This is definitely an important piece of logic. What roles are acceptable, and where?